backend/core/geolocation_service.py

```
@@ -0,0 +1,167 @@
+
+class GeolocationService:
+    """Service for IP geolocation lookups with caching."""
+
+    def __init__(self):
+        """Initialize the GeolocationService."""
+        self.logger = logging.getLogger(__name__)
+        # Using ipapi.co free tier (1000 requests/day, no API key needed)
+        self.base_url = "https://ipapi.co"
+        self.timeout = 5  # seconds
+
+    @lru_cache(maxsize=1000)
```

Ruff (0.12.2)

Use of `functools.lru_cache` or `functools.cache` on methods can lead to memory leaks

(B019)